packages/@aws-cdk/aws-servicecatalog/README.md

```diff
@@ -184,6 +184,48 @@ const product = new servicecatalog.CloudFormationProduct(this, 'Product', {
 });
 ```
 
+We can also define a `versioningStrategy` for your `ProductStack` deployment. 
+With the `Default` strategy, a new `ProductStack` is needed for each `productVersion`
+and each `productVersion` will get overwritten with the latest changes to your `ProductStack`.
+With the `RetainPreviousVersions` strategy, a local template will be retained for each `ProductStack` deployed.
+Subsequently, any `ProductStack` changes will not overwrite the current version.
```

Comment:
What does the workflow look while you are developing/testing the latest `ProductStack`? 

* Are you supposed to keep the most recent version at `Default` while iterating on it, and then flip it to `RetainPreviousVersions` as you are done?

* And then, for the next version, change it to a `fromTemplate`, add a new `fromProductStack` with `RetainPreviousVersions` and repeat?

If that's what you are thinking, best to explain that here as well, because I'm not sure that's obvious to readers.

Reply:
(As you are explaining the process to users, I hope you will then internally go "*hmm that seems like a sharp edge that I can shave off/guard against with validation*")

Reply:
The workflow would ideally have user stay with `Default` or stay with `RetainPreviousVersions` the whole way and not be constantly switching between the strategies.

If they are using `RetainPreviousVersions`. They would just update the version on their current `fromProductStack ie. from `v1` to `v2.
Then if they want the previous version to be deployed as well, they would have to add another version with `fromProductStackContext`

The end result would be the same as you explained, but I guess the workflow for the user would be to update their "current" version, then add any "previous" versions if they choose to, rather than change their "current" to use `fromProductStackContext` then add a new version.

Reply:
I think you are making the assumption that a user gets their new version of the `ProductStack` exactly right on the first attempt. Remember that if we have `RetainPreviousVersions` enabled, what happens on `synth` is:

- If no snapshot of the stack exists, we create it
- If a snapshot already exists, the current stack definition must be exactly equal to the snapshot

That means:

- If a user changes the product version name FIRST before changing the `ProductStack` and (accidentally or to check something) run `cdk synth`, their stack definition is now **locked in**. They can't change it anymore! (And they haven't even begun to make their changes yet)
  - If a user is using `cdk watch`, synthing is done continuously on their behalf. They may be synthing (and locking in their template) without even consciously taking an action.
- Even if they made the changes to the `ProductStack` *before* changing the version, and synth it, and potentially even `deploy` it to their developer account... their changes are now ALSO locked in. Iteration is effectively impossible, hope they got it right on the first try!

That means that in a realistic development workflow, people will have to switch back and forth between those versioning strategies to iterate. 

💡 Here's a tip that will last you the rest of your career: **Blog post driven development**. Look at the [AWS Blog](https://aws.amazon.com/blogs/?awsf.blog-master-category=*all&awsf.blog-master-learning-levels=*all&awsf.blog-master-industry=*all&awsf.blog-master-analytics-products=*all&awsf.blog-master-artificial-intelligence=*all&awsf.blog-master-aws-cloud-financial-management=*all&awsf.blog-master-blockchain=*all&awsf.blog-master-business-applications=*all&awsf.blog-master-compute=*all&awsf.blog-master-customer-enablement=*all&awsf.blog-master-customer-engagement=*all&awsf.blog-master-database=*all&awsf.blog-master-developer-tools=*all&awsf.blog-master-devops=*all&awsf.blog-master-end-user-computing=*all&awsf.blog-master-mobile=*all&awsf.blog-master-iot=*all&awsf.blog-master-management-governance=*all&awsf.blog-master-media-services=*all&awsf.blog-master-migration-transfer=*all&awsf.blog-master-migration-solutions=*all&awsf.blog-master-networking-content-delivery=*all&awsf.blog-master-programming-language=*all&awsf.blog-master-sector=*all&awsf.blog-master-security=*all&awsf.blog-master-storage=*all) and read a couple of posts. Notice how they all take the customer by the hand and say: "to use this feature, you first do X, then Y, then Z", and they don't skip any steps or handwave. For every feature you design henceforth, try to do the same. Pretend you are writing an AWS Blog Post, write in a friendly voice to the customer that has a particular problem or use case, telling them how awesome your feature is and that they "only" need to do X, Y and Z in that order (and don't skip anything). This will force you to be explicit about certain things you are probably handwaving away right now, and potentially cause you to scratch your chin and internally go "oh that isn't as easy as I thought after all".